### PR TITLE
[SPARK-9856][SparkR] Add expression functions into SparkR whose params are complicated

### PR DESCRIPTION
**Pull request by yu-iskw**
I added lots of Column functinos into SparkR. And I also added `rand(seed: Int)` and `randn(seed: Int)` in Scala. Since we need such APIs for R integer type.
### JIRA

[[SPARK-9856] Add expression functions into SparkR whose params are complicated - ASF JIRA](https://issues.apache.org/jira/browse/SPARK-9856)

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -682,6 +682,14 @@ object functions {
    * Generate a random column with i.i.d. samples from U[0.0, 1.0].
    *
    * @group normal_funcs
+   * @since 1.5.0
+   */
+  def rand(seed: Int): Column = Rand(Int.int2long(seed))
```
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Isn't this implicitly done by scala ?  cc @rxin -- what's our policy on using implicits like the ones in `Int` to `Long` etc ?

**Comment by shivaram** [comment]
@yu-iskw could you also check if `lint-r` is happy on your machine ?

**Comment by yu-iskw** [comment]
@shivaram Yeah, I have checked it with {{lint-r}} before sending a PR. There was no warning. Thanks!

**Comment by shivaram** [comment]
Thanks @yu-iskw -- Change looks pretty good to me and the test cases are especially good. Thanks for adding that. 

Lets see if @rxin or @davies have any thoughts on the minor Scala style thing I mentioned inline

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: i think it should just work without the int2long.

Can we mark this function as private[sql] ?

Comment (yu-iskw) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: When I tried to mark them as `private[sql]`, I couldn't invoke them from SparkR.

Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: The function needs to be visible in the `[r]` namespace. You can make it `private[spark]` and I think that should work

Comment (yu-iskw) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @shivaram hmm, although I made it `private[spark]`, it didn't work.

**Comment by yu-iskw** [comment]
@rxin it seems that `from_unixtime` and `unix_timestamp` depends on the default time zone setting.
I'm working at JST (UTC+9), but the Jenkins is at PDT(UTC-7). So those errors as follows might occur because of the time difference.
At least, we  can''t set a time zone of `from_unixtime` and `unix_timestamp`. Is that OK?

For example, `from_unixtime(1000)`
- JST(UTC+9) => "1970-01-01 09:16:40"
- PDT(UTC-7) => "1969-12-31 16:16:40"

### Jenkins Ouput
```
1. Failure (at test_sparkSQL.R#797): date functions on a DataFrame -------------
collect(select(df2, unix_timestamp(lit("2015-01-01"), "yyyy-MM-dd")))[1, 1] not equal to 1420038000
1.42e+09 - 1.42e+09 == -61200

2. Failure (at test_sparkSQL.R#802): date functions on a DataFrame -------------
collect(select(df3, from_unixtime(df3$a)))[, 1] not equal to c("1970-01-01 09:16:40", "1970-01-01 08:43:20")
2 string mismatches:
x[1]: "1970-01-01 09:16:40"
y[1]: "1969-12-31 16:16:40"

x[2]: "1970-01-01 08:43:20"
y[2]: "1969-12-31 15:43:20"
```

**Comment by davies** [comment]
You can have a test which call unix_timestamp and from_unixtime, a round trip of these two should not be affected by timezone.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (shivaram) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: hmm so the method should be invoked from the RBackendHandler which is in scope. Could you paste the stack trace that you get with `private[spark]`

@davies  Any ideas about this ?

**Comment by davies** [comment]
@yu-iskw You add this patch, then you do not need the special version of rand/randn
```
--- a/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala
+++ b/core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala
@@ -181,6 +181,7 @@ private[r] class RBackendHandler(server: RBackend)
       // Convert native parameters to Object types as args is Array[Object] here
       if (parameterType.isPrimitive) {
         parameterWrapperType = parameterType match {
+          case java.lang.Long.TYPE => classOf[java.lang.Integer]  // R does not have long
           case java.lang.Integer.TYPE => classOf[java.lang.Integer]
           case java.lang.Double.TYPE => classOf[java.lang.Double]
           case java.lang.Boolean.TYPE => classOf[java.lang.Boolean]
```

**Comment by davies** [comment]
@yu-iskw Would you mind to add doc for every function?

**Comment by yu-iskw** [comment]
@davies thank you for the comment. I'll try to do that.

I have another question about `from_unixtime`. HIve's `from_unixtime` converts the number of seconds from unix epoch refered to UTC time zone to a timestamp string. Should we follow it or not?

"Converts the number of seconds from unix epoch (1970-01-01 00:00:00 UTC) to a string representing the timestamp of that moment in the current system time zone in the format of "1970-01-01 00:00:00"."
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF

**Comment by shivaram** [comment]
@yu-iskw Can we just match the existing Scala / Python behavior in `from_unixtime` for now ? We can discuss the Hive vs. Spark things in a separate JIRA issue.

**Comment by yu-iskw** [comment]
Jenkins, test this please.

**Comment by yu-iskw** [comment]
Jenkins, retest this please.

**Comment by yu-iskw** [comment]
Jenkins, retest this please.

**Comment by yu-iskw** [comment]
Jenkins, test this please.

**Comment by yu-iskw** [comment]
Jenkins, retest this please.

**Comment by shivaram** [comment]
Thanks @yu-iskw -- LGTM. Merging this.
@davies I'll open a separate JIRA to do a pass over the R API documentation for 1.5.

**Comment by yu-iskw** [comment]
Thank you for merging it!

**Pull request closed without merging**